crates/lib/src/qpu/quilc/mod.rs

```diff
@@ -34,11 +35,15 @@ pub(crate) fn compile_program(
     quil: &str,
     isa: TargetDevice,
     client: &Qcs,
+    timeout: Option<u8>,
```

Comment:
Is it an abuse of `Option`s in Rust to use them for pseudo-optional parameters like this? I don't love it. In this case, it would be easy to think setting `None` here would set no timeout at all, but it doesn't.

At first, I opted for a `compile_with_timeout` method but then the plain `compile` method would be misleading in the same way this parameter is. In addition the plain `compile` method would fall out of use, as we want to support configuring timeouts everywhere compilation is used. 

We _could_ get rid of the default timeout of 10s, but as a user, I'd rather run into a timeout that would give me a warning rather then have the program hang until I quit it without any obvious insight into why.

Reply:
Besides the other breaking changes such as using a builder pattern or a custom parameter struct, the only other option would be an environment variable and that seems pretty inflexible, so your change seems like the best move, especially since the rust comments and `.pyi` file make it clear what the default value is.

Does setting a value of `Some(0)` make the timeout indefinite?

---

Side thought - I'm guessing the following would be overkill if not an antipattern (but a nifty thing I found [here](https://vidify.org/blog/rust-parameters/#b-with-intooptiont))

```rust
pub(crate) fn compile_program(
    // ...
    timeout: impl Into<Option<u8>>,
) -> Result<quil_rs::Program, Error> {
    // ...
    if let Some(seconds) = timeout.into() {
        request.with_timeout(seconds);
    }
    // ...
}

// allows all of these:
compile_program(/* ... */, 10);
compile_program(/* ... */, Some(10));
compile_program(/* ... */, None);
```

Reply:
`None` _should_, in effect, be no timeout... at least that's how quilc treats it. (I think it can also technically take a `false` value which indicates the request should use the _server's_ timeout setting.)

So I think the type of `client_timeout` _should_ be `Option<u8>` everywhere.

Reply:
And I'm also in favour of bumping the default timeouts to 30s here and in pyquil.

Reply:
@jselig-rigetti - That is a nifty trick. I do like that it allows passing in an int directly. I feel similarly though, not sure if it's an antipattern or not.

@notmgsk - 30s is what I was thinking for the Azure integration too, so we're aligned on setting a new default. We can make it Option<u8> all the way down. That said, if `None` is supposed to mean no timeout, how should we approach having a default of 30s? When calling `compile_program` should a `timeout: Option<u8>` of `None` mean default to 30s, or no timeout?

The latter _feels_ right, but it doesn't allow us to have a "default" timeout for the function.

Reply:
After thinking about this a little more, I think I'm just gonna wrap this in a `CompilerOptions` struct that implements `default`. It's extendable, protects us from future breaking changes, and allows us to have our 30s default without being potentially misleading to the user.

Reply:
Good idea!